pyradiosky/skymodel.py

```diff
@@ -605,31 +638,34 @@ def coherency_calc(self, deprecated_location=None):
                 errm + "value was: {al}".format(al=str(self.telescope_location))
             )
 
-        Ionly_mask = np.sum(self.stokes[1:, :, :], axis=0) == 0.0
-        NstokesI = np.sum(Ionly_mask)  # Number of unpolarized sources
+        # Select sources within the horizon only.
+        coherency_local = self.coherency_radec[..., self.above_horizon]
 
         # For unpolarized sources, there's no need to rotate the coherency matrix.
-        coherency_local = self.coherency_radec.copy()
-
-        if NstokesI < self.Ncomponents:
+        if self._n_polarized > 0:
             # If there are any polarized sources, do rotation.
-            rotation_matrix = self._calc_coherency_rotation()
 
-            polarized_sources = np.where(~Ionly_mask)[0]
+            # Indices of source components, downselected to those above the horizon,
+            # that are also polarized.
+            pol_over_hor = np.in1d(
+                np.arange(self.Ncomponents)[self.above_horizon], self._polarized
+            )
+
+            # Indices of polarized sources on the full Ncomponents array, only
+            # if they're above the horizon.
+            full_inds = [pi for pi in self._polarized if self.above_horizon[pi]]
```

Comment:
I don't understand what the difference between `pol_over_hor` and `full_inds` are. They both seem to be the indices of the polarized sources above the horizon...

Reply:
It's a subtle difference, and I'm not sure how best to name them to explain it.

The coherency matrix has been reduced to only include the sources that are above the horizon. `pol_over_hor` indexes polarized sources within that reduced array. The ra/dec arrays, however, are still length Ncomponents. `full_inds` gives the indices of sources that are polarized and above the horizon.